File: WEIRDO.py
t = int(input())
vow = ['a','e','i','o','u']

def expo(x,y):
	if y==1:
		return x
	res = expo(x,y//2)
	if y%2:
		return res*res*x
	else:
		return res*res

for _ in range(t):
	l = int(input())
	a_total,b_total,a_doc,b_doc = {},{},{},{}
	a,b = 0,0 
	for i in range(l):
		s = input().strip()
		alice = 1
		for j in range(1,len(s)):
			v,c = 0,0
			for k in range(3):
				if j-k >= 0:
					if s[j-k] in vow:
						v += 1
					else:
						c += 1
			if v<c:
				alice = 0
				break
		
		hs = [1]*27
		for j in range(len(s)):
			if alice:
				a_total[s[j]] = a_total.get(s[j],0) + 1
				if hs[ord(s[j]) - 97]:
					hs[ord(s[j]) - 97] = 0
					a_doc[s[j]] = a_doc.get(s[j],0) + 1
			else:
				b_total[s[j]] = b_total.get(s[j],0) + 1
				if hs[ord(s[j]) - 97]:
					hs[ord(s[j]) - 97] = 0
					b_doc[s[j]] = b_doc.get(s[j],0) + 1
		if alice:
			a+=1
		else:
			b+=1

	l1,l2 = [],[]
				
	for i in a_doc:
		l1.append(a_doc[i])
		l2.append((a_total[i]**a))
	for i in b_doc:
		l2.append(b_doc[i])
		l1.append((b_total[i]**b))
	l1.sort()
	l2.sort()
	
	i,j = 0,0
	sc = 1
	while i<len(l1) and j<len(l2):
		while i<len(l1) and sc <= 1:
			sc *= l1[i]
			i += 1
		while j<len(l2) and sc > 1:
			sc /= l2[j]
			j += 1
	while i<len(l1):
		sc *= l1[i]
		i += 1
	while j<len(l2):
		sc /= l2[j]
		j += 1

	score = sc
	if score > 10000000:
		print('Infinity')
	else:
		print(score)

File: WEIRDO.cpp
#include <bits/stdc++.h> 
using namespace std;

long long int expo(long long int x, long long int y){
	if(y==1) return x;
	long long int res = expo(x,y/2);
	
	if(y%2) return res*res*x;
	else return res*res;
}

int main(){
	long long int t,l,alice;
	string s;
	float score;	
	cin >> t;

	for (int q=0 ; q<t ; q++){
		cin >> l;
		map<char,long long int>a_total,b_total,a_doc,b_doc;
		long long int a=0,b = 0; 
		for(int i=0 ; i<l ; i++){
			cin >> s;
			alice = 1;
			int v,c;
			for (int j=1 ; j < s.length() ; j++){
				v = 0;
				c = 0;
				for (int k=0 ; k<3 ; k++){
					if (j-k >= 0){
						if (s[j-k]=='a' || s[j-k]=='e' || s[j-k]=='i' || s[j-k]=='o' || s[j-k]=='u') v++;
						else c++;
					}
				}
				if (v<c){
					alice = 0;
					break;
				}
			}
			bool hs[26];
			for (int z = 0; z < 26; z++)
			{
				hs[z] = true;
			}
			
			for (int j=0 ; j < s.length() ; j++){
				if (alice){
					a_total[s[j]]++;
					if(hs[int(s[j]) - 97]){
						hs[int(s[j]) - 97] = false;
						a_doc[s[j]]++;
					}
				}
				else{
					b_total[s[j]]++;
					if(hs[int(s[j]) - 97]){
						hs[int(s[j]) - 97] = false;
						b_doc[s[j]]++;
					}
				}
			}
			if(alice) a++;
			else b++;		
		}
		
		vector<long long int>l1,l2;
		for(auto pair : a_doc){
			l1.push_back(pair.second);
			l2.push_back((int)(pow(a_total[pair.first],a)+0.5));
		}
		for(auto pair : b_doc){
			l1.push_back((int)(pow(b_total[pair.first],b)+0.5));
			l2.push_back(pair.second);
		}
		sort(l1.begin(), l1.end());
		sort(l2.begin(), l2.end());
		
		float score=1;
		long int i=0,j=0;
		
		while(i<l1.size() && j<l2.size()){
			if(score<=1e15 && l1[i]<1e15){
				score *= l1[i];
				i++;
			}
			else if(score>1e15 && l1[j]<1e30){
				score /= l2[j];
				j++;
			}
		}
		while(i<l1.size()){
			score *= l1[i];
			i++;
		}
		while(j<l2.size()){
			score /= l2[j];
			j++;
		}
		
		if(score > 10000000){
			cout << "Infinity" << endl;
		}
		else {
			cout << score << endl;
		}
		
	}
	return 0;
}